applications: serial_lte_mode: add TWI AT command (#4152)
nrfconnect/sdk-nrf: applications/serial_lte_modem/doc/TWI_AT_commands.rst

.. _SLM_AT_TWI:

TWI AT commands
***************

.. contents::
   :local:
   :depth: 2

The following commands list contains TWI related AT commands.

Open TWI instance #XTWIOP
=========================

The ``#XTWIOP`` command opens TWI instance.

Set command
-----------

The set command allows you to open TWI instance.

Syntax
~~~~~~

::

   #XTWIOP=<index>

The ``<index>`` parameter accepts the following integer values:

* ``0`` - i2c0
* ``1`` - i2c1
* ``2`` - i2c2
* ``3`` - i2c3

Example
~~~~~~~

::

   AT#XTWIOP=2
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

> 

Reviewer: Do you think it'll be good to support Test to list all available TWI instance by DTS?

Author: #XTWILS is added to list all available TWI instance.

* ``0`` - i2c0
* ``1`` - i2c1
* ``2`` - i2c2
* ``3`` - i2c3

The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to write. The maximum size is 2-digit (e.g. "DE" for 0xDE)
The ``<data>`` parameter is a hexadecimal string.
  It represents the data to be written to slave device. The maximum size is 254-digit. (e.g. "DEADBEEF" for 0xDEADBEEF)

Example
~~~~~~~

::

   AT#XTWIW=2,"76","D0"
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

Read from TWI slave device #XTWIR
=================================

The ``#XTWIR`` command reads data from a TWI slave device.

Set command
-----------

The set command allows you to read data from a TWI slave device.

Syntax
~~~~~~

::

   #XTWIR=<index>,<dev_addr>,<num_read>

The ``<index>`` parameter accepts the following integer values:

* ``0`` - i2c0
* ``1`` - i2c1
* ``2`` - i2c2
* ``3`` - i2c3

The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to read. The maximum size is 2-digit (e.g. "DE" for 0xDE)
The ``<num_read>`` parameter is an unsigned 8-bit integer.

>   It represents the number of data to be read from slave device. The available range is from 0 to 255.

Reviewer: Do we know the length before the Read the value out?
I think that we should not need this param and the length is to be decided after reading out

Author: Yes we specify the length before reading the data. It is not the kind of listen and wait behavior.

Open TWI instance #XTWILS
=========================

The ``#XTWILS`` command lists all available TWI instances.

Set command
-----------

The set command allows you to list all available TWI instances.

Response syntax
~~~~~~~~~~~~~~~

::

   #XTWILS: <index>[[[,<index>],<index>],<index>]

The ``<index>`` parameter corresponds to the following TWI instances:

* ``0`` - TWI0 (``i2c0``).
* ``1`` - TWI1 (``i2c1``).
* ``2`` - TWI2 (``i2c2``).
* ``3`` - TWI3 (``i2c3``).

See :ref:`zephyr:dtbinding_nordic_nrf_twi`.

Example
~~~~~~~

::

   AT#XTWILS
   #XTWILS: 2
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

Write to TWI peripheral device #XTWIW
=====================================

The ``#XTWIW`` command writes data to a TWI peripheral device.

Set command
-----------

The set command allows you to write data to a TWI peripheral device.

Syntax
~~~~~~

::

   #XTWIW=<index>,<dev_addr>,<data>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

> See :ref:`zephyr:dtbinding_nordic_nrf_twi`.

Reviewer 2: ```suggestion
  See :ref:`zephyr:dtbinding_nordic_nrf_twi`.
```

Reviewer 2: The formatting is messing up the list, so need to add indentation.

Author: modified.

* The ``<num_read>`` parameter is an unsigned 8-bit integer.
  It represents the amount of data to read from the peripheral device.
  The available range is from 0 to 255 bytes.

Response syntax
~~~~~~~~~~~~~~~

::

   #XTWIR:
   <data>

* The ``<data>`` parameter is a hexadecimal string.
  It represents the data read from the peripheral device.

Example
~~~~~~~

::

   AT#XTWIR=2,"76",1

   #XTWIR: 61
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

Write data and read from TWI peripheral device #XTWIWR
======================================================

The ``#XTWIWR`` command writes data to a TWI peripheral device and then reads data from the device.

Set command
-----------

The set command allows you to first write data to a TWI peripheral device and then read the returned data.

Syntax
~~~~~~

::

   #XTWIW=<index>,<dev_addr>,<data>,<num_read>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

> See :ref:`zephyr:dtbinding_nordic_nrf_twi`.

Reviewer 2: ```suggestion
  See :ref:`zephyr:dtbinding_nordic_nrf_twi`.
```

Author: modified.

.. _SLM_AT_TWI:

TWI AT commands
***************

.. contents::
   :local:
   :depth: 2

The following commands list contains AT commands related to the two-wire interface (TWI).

Open TWI instance #XTWILS
=========================

The ``#XTWILS`` command lists all available TWI instances.

Set command
-----------

The set command allows you to list all available TWI instances.

Response syntax
~~~~~~~~~~~~~~~

::

   #XTWILS: <index>[[[,<index>],<index>],<index>]

The ``<index>`` parameter maps to the below TWI instance:

* ``0`` - Use TWI0 (``i2c0``).
* ``1`` - Use TWI1 (``i2c1``).
* ``2`` - Use TWI2 (``i2c2``).
* ``3`` - Use TWI3 (``i2c3``).

See :ref:`zephyr:dtbinding_nordic_nrf_twi`.

Response syntax
~~~~~~~~~~~~~~~

There is no response.

Example
~~~~~~~

::

   AT#XTWILS
   #XTWILS: 2
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

> Write to TWI slave device #XTWIW

Reviewer 2: Can we replace slave device, for example, with "peripheral device"?

Author: Yes peripheral is fine. Modified per suggestion.

=================================

The ``#XTWIW`` command writes data to a TWI slave device.

Set command
-----------

The set command allows you to write data to a TWI slave device.

Syntax
~~~~~~

::

   #XTWIW=<index>,<dev_addr>,<data>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

See :ref:`zephyr:dtbinding_nordic_nrf_twi`.

* The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to write to.
  The maximum length is 2 characters (for example, "DE" for 0xDE).
* The ``<data>`` parameter is a hexadecimal string.

>   It represents the data to be written to slave device.

Reviewer 2: ```suggestion
  It represents the data to be written to the slave device.
```

Author: Modified per suggestion.

>   The maximum length is 255 characters (for example, "DEADBEEF" for 0xDEADBEEF).

Reviewer 2: Is it correct that it is max 255 characters for the write command, but max 254 characters for the write then read command?

Author: It is actually 255. Now it is fixed.

Syntax
~~~~~~

::

   #XTWIR=<index>,<dev_addr>,<num_read>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

* The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to read from.
  The maximum length is 2 characters (for example, "DE" for 0xDE).
* The ``<num_read>`` parameter is an unsigned 8-bit integer.
  It represents the amount of data to read from the slave device.
  The available range is from 0 to 255 bytes.

Response syntax
~~~~~~~~~~~~~~~

::

   #XTWIR:
   <data>

* The ``<data>`` parameter is a hexadecimal string.
  It represents the data read from the slave device.

Example
~~~~~~~

::

   AT#XTWIR=2,"76",1

   #XTWIR: 61
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

Write data and read from TWI device #XTWIWR
===========================================

The ``#XTWIWR`` command writes data to a TWI slave device and then reads data from the device.

Set command
-----------

> The set command allows you to first write data to a TWI slave device and then read that data back.

Reviewer 2: ```suggestion
The set command allows you to first write data to a TWI slave device and then read the returned data.
```

Author: Modified per suggestion.

Syntax
~~~~~~

::

   #XTWIW=<index>,<dev_addr>,<data>,<num_read>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

See :ref:`zephyr:dtbinding_nordic_nrf_twi`.

* The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to write to.
  The maximum length is 2 characters (for example, "DE" for 0xDE).
* The ``<data>`` parameter is a hexadecimal string.

>   It represents the data to be written to slave device.

Reviewer 2: ```suggestion
  It represents the data to be written to the slave device.
```

Author: Modified per suggestion.

.. _SLM_AT_TWI:

TWI AT commands
***************

.. contents::
   :local:
   :depth: 2

The following commands list contains AT commands related to the two-wire interface (TWI).

> Open TWI instance #XTWILS

Author: for listing instances. Now the instances are opened automatically.

=========================

The ``#XTWILS`` command lists all available TWI instances.

Set command
-----------

The set command allows you to list all available TWI instances.

Response syntax
~~~~~~~~~~~~~~~

::

   #XTWILS: <index>[[[,<index>],<index>],<index>]

> The ``<index>`` parameter maps to the below TWI instance:

Author: Modified per suggestion.

* ``0`` - Use TWI0 (``i2c0``).
* ``1`` - Use TWI1 (``i2c1``).
* ``2`` - Use TWI2 (``i2c2``).

> * ``3`` - Use TWI3 (``i2c3``).

Author: Modified per suggestion.

See :ref:`zephyr:nordic_nrf_twi`.

Response syntax
~~~~~~~~~~~~~~~

> There is no response.

Author: Redundant... it is now deleted.

  It represents the data to be written to slave device.
  The maximum length is 254 characters (for example, "DEADBEEF" for 0xDEADBEEF).

Response syntax
~~~~~~~~~~~~~~~

There is no response.

Example
~~~~~~~

::

   AT#XTWIW=2,"76","D0"
   OK

Read command
------------

The read command is not supported.

Test command
------------

The test command is not supported.

Read from TWI slave device #XTWIR
=================================

The ``#XTWIR`` command reads data from a TWI slave device.

Set command
-----------

The set command allows you to read data from a TWI slave device.

Syntax
~~~~~~

::

   #XTWIR=<index>,<dev_addr>,<num_read>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

  See :ref:`zephyr:nordic_nrf_twi`.

* The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to read from.
  The maximum length is 2 characters (for example, "DE" for 0xDE).
* The ``<num_read>`` parameter is an unsigned 8-bit integer.
  It represents the amount of data to read from the slave device.

>   The available range is from 0 to 255 bytes.

Reviewer 2: Is bytes correct here?

Author: yes, correct.

Syntax
~~~~~~

::

   #XTWIW=<index>,<dev_addr>,<data>,<num_read>

* The ``<index>`` parameter accepts the following integer values:

  * ``0`` - Use TWI0 (``i2c0``).
  * ``1`` - Use TWI1 (``i2c1``).
  * ``2`` - Use TWI2 (``i2c2``).
  * ``3`` - Use TWI3 (``i2c3``).

  See :ref:`zephyr:nordic_nrf_twi`.

* The ``<dev_addr>`` parameter is a hexadecimal string.
  It represents the slave device address to write to.
  The maximum length is 2 characters (for example, "DE" for 0xDE).
* The ``<data>`` parameter is a hexadecimal string.
  It represents the data to be written to slave device.
  The maximum length is 254 characters (for example, "DEADBEEF" for 0xDEADBEEF).
* The ``<num_read>`` parameter is an unsigned 8-bit integer.
  It represents the amount of data to read from the slave device.
  The available range is from 0 to 255 bytes.

Response syntax
~~~~~~~~~~~~~~~

::

   #XTWIWR: <size>
   <data>

* The ``<size>`` value is an unsigned integer.
  It represents the actual number of bytes read from the slave device.
* The ``<data>`` parameter is a hexadecimal string.
  It represents the data read from the slave device.

Example
~~~~~~~

::

   AT#XTWIWR=2,"76","D0",1

   #XTWIWR: 2

>    61

Reviewer 2: Is this the correct return value? I though it was hexadecimal (which might still be correct, I just don't understand how we get there from "D0". ;-) ).

Author: Yes, it (61) is correct return value for hexadecimal value 0x61.

AT#XTWIWR wraps the [i2c_write_read](https://docs.zephyrproject.org/apidoc/latest/group__i2c__interface_ga1273a9f8bdb002e0d6ece3a8c893a709.html) function that reading data from a register of a i2c slave device.
As the API description said it is a common operation "this is what I want", "now give it to me".

The example **AT#XTWIWR=2,"76","D0",1** reads the CHIP ID register of BME680 sensor on Thingy:91. The return value 61 is the CHIP ID of BME680.

As a reference, you can find the definition of registers here:
https://github.com/nrfconnect/sdk-zephyr/blob/master/drivers/sensor/bme680/bme680.h#L36
And the BME680 chip id definition:
https://github.com/nrfconnect/sdk-zephyr/blob/master/drivers/sensor/bme680/bme680.h#L14
